File: src/Cxx/Widgets/CompassWidget.cxx
#include <vtkAnnotatedCubeActor.h>
#include <vtkCamera.h>
#include <vtkCommand.h>
#include <vtkCompassRepresentation.h>
#include <vtkCompassWidget.h>
#include <vtkNamedColors.h>
#include <vtkNew.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>

#if VTK_VERSION_NUMBER >= 90220220831ULL
#define VTK_HAS_IMPROVED_COMPASSWIDGETREPRESENTATION 1
#endif

class vtkICWValueChangedCallback : public vtkCommand
{
public:
  static vtkICWValueChangedCallback* New()
  {
    return new vtkICWValueChangedCallback();
  }
  virtual void Execute(vtkObject* caller, unsigned long vtkNotUsed(eventId),
                       void* vtkNotUsed(callData))
  {
    vtkCompassWidget* widget = vtkCompassWidget::SafeDownCast(caller);
    vtkCamera* camera = widget->GetCurrentRenderer()->GetActiveCamera();

    // calculate new camera position from compass widget parameters
    double distance = widget->GetDistance();
    double tilt = widget->GetTilt();
    double heading = widget->GetHeading();

    double pos[3] = {0, 0, 0};
    pos[0] = distance * cos(vtkMath::RadiansFromDegrees(heading)) *
        cos(vtkMath::RadiansFromDegrees(tilt));
    pos[1] = distance * sin(vtkMath::RadiansFromDegrees(heading)) *
        cos(vtkMath::RadiansFromDegrees(tilt));
    pos[2] = distance * sin(vtkMath::RadiansFromDegrees(tilt));

    camera->SetPosition(pos);
    camera->SetFocalPoint(0, 0, 0);
    camera->SetViewUp(0, 0, 1);
    camera->SetClippingRange(0.1, distance + 1);

    widget->GetCurrentRenderer()->Render();
  }
  vtkICWValueChangedCallback()
  {
  }
};

int main(int, char*[])
{
  vtkNew<vtkNamedColors> colors;

  // a cube with text on its faces
  vtkNew<vtkAnnotatedCubeActor> actor;
  actor->GetTextEdgesProperty()->SetColor(
      colors->GetColor3d("Black").GetData());
  actor->GetCubeProperty()->SetColor(colors->GetColor3d("PeachPuff").GetData());

  // a renderer and render window
  vtkNew<vtkRenderer> renderer;
  vtkNew<vtkRenderWindow> renderWindow;
  renderWindow->AddRenderer(renderer);

  // an interactor
  vtkNew<vtkRenderWindowInteractor> renderWindowInteractor;
  renderWindowInteractor->SetRenderWindow(renderWindow);

  // create the widget and its representation
  vtkNew<vtkCompassRepresentation> compassRepresentation;
#ifdef VTK_HAS_IMPROVED_COMPASSWIDGETREPRESENTATION
  compassRepresentation->SetMinimumDistance(2);
  compassRepresentation->SetMaximumDistance(10);
#endif

  vtkNew<vtkCompassWidget> compassWidget;
  compassWidget->SetInteractor(renderWindowInteractor);
  compassWidget->SetRepresentation(compassRepresentation);
  compassWidget->SetDistance(5.0);

  // create the callback
  vtkNew<vtkICWValueChangedCallback> valueChangedCallback;
  compassWidget->AddObserver(vtkCommand::WidgetValueChangedEvent,
                             valueChangedCallback);

  // add the actors to the scene
  renderer->AddActor(actor);
  renderer->SetBackground(colors->GetColor3d("MidnightBlue").GetData());

  renderWindow->SetSize(640, 480);
  renderWindow->SetWindowName("CompassWidget");

  renderWindow->Render();
  compassWidget->EnabledOn();

  // no interactor style - camera is moved by widget callback
  renderWindowInteractor->SetInteractorStyle(nullptr);

  // set camera to initial position
  compassWidget->InvokeEvent(vtkCommand::WidgetValueChangedEvent);

  // begin interaction
  renderWindowInteractor->Start();

  return EXIT_SUCCESS;
}

File: src/Python/Widgets/CompassWidget.py
#!/usr/bin/env python3

import math
# noinspection PyUnresolvedReferences
import vtkmodules.vtkInteractionStyle
# noinspection PyUnresolvedReferences
import vtkmodules.vtkRenderingOpenGL2
from vtkmodules.vtkCommonCore import (
    VTK_VERSION_NUMBER,
    vtkCommand,
    vtkMath,
    vtkVersion
)
from vtkmodules.vtkCommonColor import vtkNamedColors
from vtkmodules.vtkRenderingAnnotation import vtkAnnotatedCubeActor
from vtkmodules.vtkInteractionStyle import vtkInteractorStyleTrackballCamera
from vtkmodules.vtkRenderingCore import (
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkRenderer
)

try:
    from vtkmodules.vtkInteractionWidgets import (
        vtkCompassRepresentation,
        vtkCompassWidget
    )
except:
    from vtkmodules.vtkGeovisCore import (
        vtkCompassRepresentation,
        vtkCompassWidget
    )

def vtk_version_ok(major, minor, build):
    """
    Check the VTK version.

    :param major: Major version.
    :param minor: Minor version.
    :param build: Build version.
    :return: True if the requested VTK version is greater or equal to the actual VTK version.
    """
    needed_version = 10000000000 * int(major) \
                     + 100000000 * int(minor) \
                     + int(build)
    try:
        vtk_version_number = VTK_VERSION_NUMBER
    except AttributeError:
        # Expand component-wise comparisons for VTK versions < 8.90.
        ver = vtkVersion()
        vtk_version_number = 10000000000 * ver.GetVTKMajorVersion() \
                             + 100000000 * ver.GetVTKMinorVersion() \
                             + ver.GetVTKBuildVersion()
    if vtk_version_number >= needed_version:
        return True
    else:
        return False

def CompassWidgetValueChangedCallback(widget, event):
    """
    Callback to set the camera position according to the position parameters given by the vtkCompassWidget.
    """
    try:
        camera = widget.GetCurrentRenderer().GetActiveCamera()
    except AttributeError:
        return
    
    # calculate new camera position from compass widget parameters
    distance = widget.GetDistance()
    tilt = widget.GetTilt()
    heading = widget.GetHeading()
    
    x = distance * math.cos(vtkMath.RadiansFromDegrees(heading)) * math.cos(vtkMath.RadiansFromDegrees(tilt))
    y = distance * math.sin(vtkMath.RadiansFromDegrees(heading)) * math.cos(vtkMath.RadiansFromDegrees(tilt))
    z = distance * math.sin(vtkMath.RadiansFromDegrees(tilt))
    
    camera.SetPosition(x, y, z)
    camera.SetFocalPoint(0, 0, 0)
    camera.SetViewUp(0, 0, 1)
    camera.SetClippingRange(0.1, distance + 1)
    
    widget.GetCurrentRenderer().Render()



def main():
    use_improved_compass_widget = vtk_version_ok(9, 2, 20220831)
    
    colors = vtkNamedColors()

    actor = vtkAnnotatedCubeActor()
    actor.GetCubeProperty().SetColor(colors.GetColor3d('PeachPuff'))

    # a renderer and render window
    renderer = vtkRenderer()
    renderWindow = vtkRenderWindow()
    renderWindow.AddRenderer(renderer)

    # an interactor
    renderWindowInteractor = vtkRenderWindowInteractor()
    renderWindowInteractor.SetRenderWindow(renderWindow)

    # create the widget and its representation
    compassRepresentation = vtkCompassRepresentation()

    compassWidget = vtkCompassWidget()
    compassWidget.SetInteractor(renderWindowInteractor)
    compassWidget.SetRepresentation(compassRepresentation)
    
    # add a callback to update the camera position on vtkCommand::WidgetValueChangedEvent
    compassWidget.AddObserver(vtkCommand.WidgetValueChangedEvent, CompassWidgetValueChangedCallback)

    if use_improved_compass_widget:
        compassRepresentation.SetMinimumDistance(2)
        compassRepresentation.SetMaximumDistance(10)
        compassWidget.SetDistance(5)
        compassWidget.SetTiltSpeed(45)
        compassWidget.SetDistanceSpeed(2)
        
    # add the actors to the scene
    renderer.AddActor(actor)
    renderer.SetBackground(colors.GetColor3d('MidnightBlue'))

    renderWindow.SetSize(640, 480)
    renderWindow.SetWindowName('CompassWidget')

    renderWindow.Render()
    compassWidget.EnabledOn()

    if use_improved_compass_widget:
        # no interactor style - camera is moved by widget callback
        renderWindowInteractor.SetInteractorStyle(None)
        # set camera to initial position
        compassWidget.InvokeEvent(vtkCommand.WidgetValueChangedEvent)

    # begin interaction
    renderWindowInteractor.Start()


if __name__ == '__main__':
    main()
